add option to store trees without data

rened opened the pull request:
The aim is to construct a tree on a very large data set and keep the actual data outside of the tree. This allows to later move the data to disc, mmap the array, and use the tree with the data residing on disk.

This PR adds a type DataFreeTree  (I am open for alternative names...) and a keyword argument `storedata::Bool` which controls whether the data is stored with the tree.

If `storedata==true` everything is just as before this PR. If `storedata==false`, `reorder` is set to `false`, the tree gets constructed, but stored with an empty matrix and the returned tree is wrapped in a `DataLessTree` type.

Before such a tree can be used, it has to be linked with data using `injectdata(datalesstree, data)`, which returns a standard `KDTree` et al for use as usual.

rened commented:
I don't know why the win32 tests would be failing while everything else passes on 0.4 and 0.5...

KristofferC commented on src/datafreetree.jl:
```diff
@@ -0,0 +1,14 @@
+type DataFreeTree{T <: AbstractFloat, M <: Metric}
+    size::Tuple{Int64,Int64}
```
Isn't this what is causing the test failures? On 32 bit `size(data)` will be a `Tupel{Int32, Int32}`. Shouldn't you just have Int here?

rened replied:
doh! of course, thanks!

KristofferC commented:
Thanks this is very nice to have. I've gotten some questions on the mail about saving trees to disk and this would make that much nicer. Some questions and comments.

* This will give a type instability in the tree constructors. Do you think this will be a problem? For the type construction it is fine but I am a little bit worried that this instability leaks out into other parts of a users program.

* The type of the `data` field is right now a `Matrix`. Is an mmapped array also of the concrete type `Matrix`? My guess is no. Maybe the tree should be parametrisized on an `AbstractMatrix` instead.  What do you think?

KristofferC commented:
What do you think about computing some sort of checksum (instead of just dimensions) to see that you are actually loading the correct data set. I am thinking the case where a user has many data sets of the same size and accidentally loads the wrong one into a tree. Could cause silent errors.

rened commented:
Regarding mmapped array: The users tells `mmap` which type the resulting array should have, so that should work just fine:
```jl
julia> a = Mmap.mmap("asdf", Array{UInt8,2}, (2,2))
2x2 Array{UInt8,2}:
 0x61  0x64
 0x73  0x66
```

rened commented:
Checksum: very good point, will add that!

Regaring the type instabilites, I will into that.

rened commented:
I could make it so that I revert the changes to the 3 constructors and do it along these lines instead:
```jl
t = DataFreeTree(KDTree, args...; kargs...)
```
Inside of that KDTree(args...; kargs..., reorder = false) gets called. That should avoid any type instabilities?

KristofferC commented:
Yes, using a completely different constructor for the `DataFreeTree` would get rid of the type instabilities of the normal tree constructors.

It would be nice if the `DataFreeTree` could be parametrisized on the tree that it stores so that the `injectdata` method could be type stable but I think that might be impossible without "triangular dispatch"?

KristofferC commented:
Maybe we could add a test (`@test_throws`) that checks that the failure modes (wrong dimension, wrong hash) are properly detected?

rened commented:
The only way I see to have everything type stable is to have 3 different `DataFreeXXTree` types? Not pretty, but given that it is only 3 and this library is meant to be a very high performance building block I would be fine with adding them.

KristofferC commented:
I thought about this a little and the type instability should be fine (at least until we figure out something better) as long as we point it out.

It's a bit of a shame that this automatically turns of reordering because my guess is that for very large data, running with reorderering off, will mean you basically fetch from disk every time you need a point.

We could remove the `reorder = storedata ? reorder : false` bit and explicitly say in the documentation that if reordering is on the data that needs to be saved is the one in the tree. We should probably then also export a `getdata(::NNTree)` function.

Anyway, all of this could be added later. This looks good to me. It would be nice to have some info about it in the readme. Best would probably be after the examples of creating trees. Also some small docstrings (similar to the ones existing) on `injectdata` + `DataFreeTree` would be nice.

Thanks for doing this. I never realized how big problems some people have. A guy emailed me and said he wanted to run nearest neighbor stuff on a 300 GB data set... That would be cool to support with shared memory arrays.

rened commented:
Thanks for your input!

I'll add a way to either have `reorder==false` (like now), or alternatively have the reordered data returned as well (which can then be saved to disk and reused).

I'll add docs and benchmarks. I am curious how fast we can get that 300GB indexing ;-)

rened commented:
@KristofferC I've added the option to provide a custom reorder buffer, meaning one can reorder a 100GB on disk data set into another 100GB on disk file if so desired. Also added docs to code and README.

I fear I might have introduced a type instability, as a `@profile` of `knn` shows a lot of inference going on... I still have to take a look at that - do you spot something that might have caused this?

KristofferC commented:
Thanks for updating. I'll look over it and see if I can find the problem.

KristofferC commented on README.md:
```diff
@@ -44,6 +44,37 @@ balltree = BallTree(data, Minkowski(3.5); reorder = false)
+
+By default, the trees store a copy of the `data` provided during construction. This is problemtic in case you want to work on data sets which are larger than available memory, thus wanting to `mmap` the data or want to store the data in a different place, outside the tree.
+
+`DataFreeTree` can be used to strip a constructed tree of it's data field and re-link it with that data at a later stage. An example of using a large on-disk data set looks like this:
```
*its

KristofferC commented on README.md:
*together

KristofferC commented on README.md:
````diff
@@ -44,6 +44,37 @@ balltree = BallTree(data, Minkowski(3.5); reorder = false)
+To perform look-ups, `dftree` is re-linked to the underlying data:
+
+```jl
+tree = injectdata(DataFreeTree(KDTree, data), data)  # yields a KDTree
````
Maybe just:

```jl
tree = injectdata(dftree, data)  # yields a KDTree
```
?

KristofferC commented on README.md:
`reorderbuf` or change the variable above to `reorderbuffer`

KristofferC commented:
I see now there is a type instability when the type of the input data is not the same as the parametrisation of the tree. Else I don't see a problem with type inference. Since `injectdata` is type unstable you will be penalized each time you access the tree in that same function so maybe that is what you are seeing?

rened commented:
Thanks for your input! I will play a bit more with this and see whether there is a performance regression somewhere.

KristofferC commented:
Do you know if there is anyway to do the equivalent of 

```jl
map(Float32, rand(2,3))
```

in a type stable way?

KristofferC commented:
Oh and btw, in the benchmark script, instead of:

```jl
r = @elapsed for i = 1:1000
            ind = rand(1:n)
            knn(t, data[:,ind], 3)[2]
        end
```
a better way would be

```jl
@elapsed r = knn(t, data[rand(1:n, 1000) ,ind], 3)[2]
```

rened commented:
hm, only via type assertion?
```jl
map(Float32, rand(2,3))::Matrix{Float32}
```

KristofferC commented:
I need it https://github.com/KristofferC/NearestNeighbors.jl/blob/master/src/knn.jl#L43 and I don't know what type of AbstractArray I will have there...

rened commented:
But you do have `T`, no? `map(T, rand(2,3))::Matrix{T}` should work?

KristofferC commented:
Above I meant `@elapsed r = knn(t, data[:, rand(1:n, 1000)], 3)[2]`

KristofferC commented:
Yes but it might also be a vector. Guess I could split it up into two functions but feels a bit awkward.

rened commented:
Now I see. Well, yes, two methods might be easiest.

rened commented:
Fixed the benchmark - reordering yields up to 2.3x speedup, using 10mio 10dimensional samples. The difference will get much bigger when the data set is larger and on a spinning disc.

rened commented:
gotta go for now - any other things I could improve tomorrow?

KristofferC commented:
I made another version of the benchmark script here so it uses the Benchmark package: https://gist.github.com/KristofferC/d0c4094b1840a3847463

We could also put the benchmark script in a new top level folder called "benchmark". After that this looks good to go. This is my first time someone has really contributed to a package I made and it feels very rewarding. Thank you :)

KristofferC commented:
Oh there is one more last thing we must fix and it is related to https://github.com/JuliaGeometry/KDTrees.jl/issues/17. Here: https://github.com/KristofferC/NearestNeighbors.jl/blob/master/src/knn.jl#L30 we translate back from the reordered indices to the original ones that existed in the input data.

However, when you inject the reordered data you presumable want to get the indices to the reordered data and not to the original data.

We could either just always return the indices to the tree.data and document it or we could add an extra function parameter. Not sure what is best.

rened commented:
Yes, collaborating on github is just great :-)  Building nice stuff out of mutual interest in methods and efficient code - perfect :-)

Regarding the returned indices, I see the need for providing the indices to both the original and the reordered data. Working on that now.

rened commented:
@KristofferC, added an option for choosing the indexing and updated the benchmark - this should be good to go, I think.

KristofferC commented:
LGTM, thanks for your contibution! We should think about tagging a new release soon to include this I think.

KristofferC commented:
Pushed a PR to tag a new version on metadata.

rened commented:
Great, thanks!
